Add support for fragmenting of DTLS handshake messages

mpg opened the pull request:
## Description

Add support for fragmenting of DTLS handshake messages based on maximum fragment length and/or an explicit MTU setting.

Supersedes and continues  https://github.com/ARMmbed/mbedtls/pull/1879 
Partially fixes https://github.com/ARMmbed/mbedtls/issues/387 (for DTLS only).
**Note to gatekeeper:** please close internal ref IOTSSL-592 when merging this.

Based on #1942 and merging #1915 and #1955 along the way as they're all necessary for tests to pass here.
Reviewers, please only review new commits in this branch.
**Note to gatekeeper:** you probably want to merge #1942, #1915 and #1955 before this one for clarity.

## Status

**READY**

## Requires Backporting

NO

## Migrations

Compatible API extensions

ABI change in libmbedtls (added `mtu` field to `stuct mbedtls_ssl_config`).

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -6881,19 +7071,59 @@ size_t mbedtls_ssl_get_max_frag_len( const mbedtls_ssl_context *ssl )
-     * Check if a smaller max length was negotiated
-     */
+    /* Check if a smaller max length was negotiated */
     if( ssl->session_out != NULL &&
```
Some tests fail here because of #1941.

mpg replied:
I believe this was fixed by basing this PR on #1942.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -958,6 +958,10 @@ struct mbedtls_ssl_config
 #endif
 
+#if defined(MBEDTLS_SSL_PROTO_DTLS)
+    uint16_t mtu;                   /*!< path mtu, used to fragment outoing messages */
```
Typo

mpg replied:
Thanks, fixed.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2423,6 +2427,35 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
+ *
+ * \note           This only controls the size of the packet we send.
+ *                 Client-side, you can request the server to use smaller
+ *                 records with \c mbedtls_conf_max_frag_len().
```
I think the Doxygen links are automatically inserted when using the `func_name()` syntax, without `\c`.

mpg replied:
Autolinks also work with the `\c`, so I'll keep it for consistency with the rest of the file. In that case, the link didn't work because the function name was wrong - missing the `ssl_` part.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2423,6 +2427,35 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
+ *                 Client-side, you can request the server to use smaller
+ *                 records with \c mbedtls_conf_max_frag_len().
+ *
+ * \note           If both a MTU and a maximum fragment length have been
```
This slightly reads as if MTU and MFL would limit the same thing, but MTU limits the maximum size of (the list of) protected records (within the current datagram), while MFL limits the maximum payload size for each single record (as far as I understand).

mpg replied:
I'll try to improve the wording here.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2423,6 +2427,35 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
+ *                 configured (or negotiated with the peer), the lower limit
+ *                 is used.
+ *
+ * \note           Values larger than \c MBEDTLS_SSL_OUT_CONTENT_LEN have no
```
I would rather expect `MBEDTLS_SSL_OUT_BUFFER_LEN` here, because `MBEDTLS_SSL_OUT_CONTENT_LEN` is about the unprotected payload.

mpg replied:
I'm rephrasing that too.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2709,19 +2757,43 @@ int mbedtls_ssl_get_record_expansion( const mbedtls_ssl_context *ssl );
+/**
+ * \brief          Return the current maximum outgoing record payload in bytes.
+ *                 This takes into account the config.h setting \c
+ *                 MBEDTLS_SSL_OUT_CONTENT_LEN, the configured and negotiated
```
I think Doxygen links are inserted when using `#MBEDTLS_SSL_OUT_CONTENT_LEN`.

mpg replied:
Thanks, fixing.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -2927,43 +3013,75 @@ void mbedtls_ssl_send_flight_completed( mbedtls_ssl_context *ssl )
+ *  - ssl->out_msg[0]: the handshake type (ClientHello, ServerHello, etc)
+ *  - ssl->out_msg + 4: the handshake message body
+ *
+ * Ouputs, ie state before passing to flight_append() or write_record():
```
Typo

mpg replied:
Thanks, fixing.

hanno-arm commented:
@mpg How do we resolve the duplicated fix for #1941? Do you want to rebase this PR on top of #1942 and remove the respective commits from here? #1942 has backports #1944 and #1945.

hanno-arm commented:
@mpg How shall we resolve the duplicated fix for #1913 and #1914? Do you want to rebase this PR on top of #1915?

mpg commented:
@hanno-arm I've solved that by merging #1915 into my branch in place of the commits duplicating it. I suggest you rebase #1918 one last time over the current state of this branch, then I'll refrain from rewriting history in this branch and will only add new commits from now on. That way we should avoid duplicated work, as well as near-duplicated commits in the overall git history after merging all those branches into development.

mpg commented:
@hanno-arm @k-stachowiak This branch is now ready for review, as I don't have any further work planned here except as a reaction to your future feedback.

@hanno-arm Please check which parts of the feedback you left on #1879 are still relevant and raise it again here when appropriate.

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -41,6 +41,28 @@ G_SRV="$GNUTLS_SERV --x509certfile data_files/server5.crt --x509keyfile data_fil
+    O_LEGACY_CLI=false
+fi
+
+if [ -n "${GNUTLS_NEXT_SERV}" ]; then
```
This makes `ssl-opt.sh` fail for me with `parameter not set` if `GNUTLS_NEXT_SERV` is not set by the environment. Similarly for `GNUTLS_NEXT_CLI`.

hanno-arm replied:
Using `${GNUTLS_NEXT_SERV:-}`` and `{GNUTLS_NEXT_CLI:-}` instead might do.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -2822,20 +2822,49 @@ static void ssl_swap_epochs( mbedtls_ssl_context *ssl )
+int mbedtls_ssl_flight_transmit( mbedtls_ssl_context *ssl )
 {
-    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> mbedtls_ssl_resend" ) );
+    const int ret_payload = mbedtls_ssl_get_max_out_record_payload( ssl );
```
Blocker: This value depends on the activated outgoing transform, which changes during flight transmission. While during the initial handshake, this rather leads to an underapproximation of the available space for the payload, during a renegotiation it might be a strict overapproximation, leading to packets of size larger than the configured MTU to be emitted. I tested this with a connection initially using AES and then switching to ChaChaPoly, and using long session ticket that needs fragmentation. In this case, the available space for the SessionTicket is judged by the maximum payload size in the ChaChaPoly suite, but it's actually encrypted using the AES transform.

This can be fixed by re-calculating `ret_payload` and `max_hs_fragment_len` in each iteration, after a potential swapping of transforms.

mpg replied:
Nice catch! Fixing.

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -4877,6 +4940,657 @@ run_test    "DTLS reassembly: fragmentation, nbio (openssl server)" \
+requires_config_enabled MBEDTLS_RSA_C
+requires_config_enabled MBEDTLS_ECDSA_C
+run_test    "DTLS fragmenting: both (MTU)" \
+            "$P_SRV dtls=1 debug_level=2 auth_mode=required \
```
Can we add an MTU proxy here, too?

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -5374,6 +5439,158 @@ run_test    "DTLS fragmenting: openssl client, DTLS 1.0" \
+             crt_file=data_files/server7_int-ca.crt \
+             key_file=data_files/server7.key \
+             mtu=512 force_version=dtls1" \
+            "$O_LEGACY_CLI -nbio -dtls1" \
```
This should be guarded by `requires_openssl_legacy()`

hanno-arm commented:
@mpg Added commits bc73e4a and 4a9d006 in #1918 fixing the aforementioned issues in `ssl-opt.sh` in the attempt to get clean tests. If you like, you can cherry-pick these here and I'll rebase #1918 and #1951.

hanno-arm commented:
The test [DTLS proxy: 3d, openssl server, fragmentation, nbio](https://github.com/ARMmbed/mbedtls/blob/38110dfc0e2f59604f6d39093471ae790323c5dc/tests/ssl-opt.sh#L5884) fails for `seed=10` because of the OpenSSL issue https://github.com/openssl/openssl/issues/6902.

mpg commented:
@hanno-arm Thanks for your review! I believe I've addressed all of your comments in the recently added commits. Feel free to review again.

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -4877,6 +4940,659 @@ run_test    "DTLS reassembly: fragmentation, nbio (openssl server)" \
+            0 \
+            -s "fragmenting handshake message"
+
+requires_config_enabled MBEDTLS_SSL_PROTO_DTLS
```
This needs `requires_gnutls`.

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -4877,6 +4940,659 @@ run_test    "DTLS reassembly: fragmentation, nbio (openssl server)" \
+requires_config_enabled MBEDTLS_RSA_C
+requires_config_enabled MBEDTLS_ECDSA_C
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_1
+run_test    "DTLS fragmenting: openssl server, DTLS 1.0" \
```
This needs `requires_gnutls`.

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -4877,6 +4940,659 @@ run_test    "DTLS reassembly: fragmentation, nbio (openssl server)" \
+            -c "fragmenting handshake message" \
+            -C "error"
+
+requires_config_enabled MBEDTLS_SSL_PROTO_DTLS
```
This needs `requires_gnutls`.

mpg replied:
Fixing, as well as the others.

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -4877,6 +4940,659 @@ run_test    "DTLS reassembly: fragmentation, nbio (openssl server)" \
+#
+# here and below we just want to test that the we fragment in a way that
+# pleases other implementations, so we don't need the peer to fragment
+requires_config_enabled MBEDTLS_SSL_PROTO_DTLS
```
This needs `requires_gnutls`.

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -4877,6 +4940,659 @@ run_test    "DTLS reassembly: fragmentation, nbio (openssl server)" \
+requires_config_enabled MBEDTLS_SSL_PROTO_DTLS
+requires_config_enabled MBEDTLS_RSA_C
+requires_config_enabled MBEDTLS_ECDSA_C
+run_test    "DTLS fragmenting: proxy MTU, resumed handshake" \
```
I see spurious failures here because sometimes messages seem to get lost in the UDP proxy during the reconnect, leading to a resend which makes the test fail.

hanno-arm replied:
Trying to address this with dc52ebd.

hanno-arm commented on ChangeLog:
```diff
@@ -2,6 +2,12 @@ mbed TLS ChangeLog (Sorted per branch, date)
+Features
+   * Add support for fragmentation of outgoing DTLS handshake messages. This
+     is controlled by the maximum fragment length as set locally or negotiated
+     with the peer, as well as new per-connection MTU option, set using
```
Minor: Is there a `by` missing in the last part? `... with the peer, and by the new per-connection MTU option, ...`?

mpg replied:
I fee like it could work both ways, but I think it's indeed clearer with a "by" in the second part. Adding it.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -1374,6 +1378,46 @@ void mbedtls_ssl_set_bio( mbedtls_ssl_context *ssl,
+ *                 example when a PMTU estimate becomes available from other
+ *                 sources, such as lower (or higher) protocol layers.
+ *
+ * \note           This only controls the size of the packets we send.
```
Suggestion: I'd find it slightly clearer to say explicitly what it does _not_ control: `This only controls the size of the packets we send, but does not restrict the size of packages we're willing to receive.`

mpg replied:
Good point. Adding that.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -1374,6 +1378,46 @@ void mbedtls_ssl_set_bio( mbedtls_ssl_context *ssl,
+ * \note           This can only be used to decrease the maximum size
+ *                 of datagrams sent. It cannot be used to increase the
+ *                 maximum size of records over the limit set by
+ *                 #MBEDTLS_SSL_OUT_CONTENT_LEN.
```
I think I disagree here. `MBEDTLS_SSL_OUT_BUFFER_LEN` is an internal macro (defined in `ssl_internal.h`), so I'm reluctant to mention it in user-facing documentation. I took care to mention "size of the records" rather than "of the datagrams" in the beginning of the sentence for consistency with the macro I'm mentioning. I'm now adding something to the previous sentence to make the connection between datagram size and record size more apparent, but I really think the only macro we should mention here is `MBEDTLS_SSL_OUT_CONTENT_LEN`.

hanno-arm replied:
Sorry for being picky here, but what I'm missing is the distinction between plain and ciphertext. What about saying `This can only be used to decrease the maximum size of datagrams (hence protected records, as those cannot span multiple datagrams). It cannot be used to increase the maximum size of protected records over the limit implied by the maximum record payload size #MBEDTLS_SSL_OUT_CONTENT_LEN`.?

mpg replied:
I think what you are suggesting is correct, but heavier that necessary. (Also, the text in parenthesis could be slightly misleading: unprotected records can't span multiple datagrams either.) The MTU is a limit on the size of protected records, which is practice is then converted to a limit on the size of unprotected records (this is already hinted at in the previous note). By switching to that view, the second part of the sentence is slightly more straightforward.

mpg replied:
OK, I just pushed a new version trying to clarify this further.

mpg commented:
@hanno-arm Thanks for your second review! I think I've addressed all your comments, please review one more time!

@k-stachowiak Please review when you can, or let me know if you won't be able to review this week so that I can find another second reviewer.

mpg commented:
Note: the Travis CI failure is due to the session resumption test, where the client is resending while it isn't expected to.

k-stachowiak commented:
@mpg I am just starting the review.

k-stachowiak commented on library/ssl_cli.c:
```diff
@@ -2721,7 +2730,7 @@ static int ssl_parse_certificate_request( mbedtls_ssl_context *ssl )
      * regardless of the actual code path.
      */
-    if( ssl->in_hslen <= mbedtls_ssl_hs_hdr_len( ssl ) + 3 + n )
+    if( ssl->in_hslen <= mbedtls_ssl_hs_hdr_len( ssl ) + 2 + n )
```
This change contradicts the comment above. I think it should be updated.

k-stachowiak commented on library/ssl_tls.c:
```diff
@@ -2855,21 +2871,91 @@ int mbedtls_ssl_resend( mbedtls_ssl_context *ssl )
+        else
+        {
+            const int ret_payload = mbedtls_ssl_get_max_out_record_payload( ssl );
+            const size_t max_record_payload = (size_t) ret_payload;
```
`max_record_payload` seems redundant (I think I have already commented that, but it may have got lost during some history rewriting).

hanno-arm replied:
I disagree - I think it's easier to read this way when distinguishing between the error code and the size value extracted from it, and the compiler will optimize it anyway. Also, in the follow-up PR #1918, this slightly changes, and the size extracted from the return value is needed more than once.

mpg replied:
Of well, not removing after all, see above.

hanno-arm commented:
@mpg Actually, now we have run into a failure during the initial `ClientHello, HelloVerifyRequest, ClientHello` exchange: https://api.travis-ci.org/v3/job/418719607/log.txt I think we should remove all `-C "resend"` guards for now.

mpg commented on tests/ssl-opt.sh:
```diff
@@ -4877,6 +4940,667 @@ run_test    "DTLS reassembly: fragmentation, nbio (openssl server)" \
+             key_file=data_files/server8.key \
+             mtu=1450 reconnect=1 reco_delay=1" \
+            0 \
+            -S "resend" \
```
I'm doing that, but instead of doing it just for this test, I'm doing it for all proxy MTU tests, because if your reasoning is correct (which I think it is), then the same kind of spurious failure can happen in any such test (even though for some reason it seems more likely to happen here).

hanno-arm replied:
@mpg It can happen for any DTLS test, even without proxy -- in fact, see the link below, it did happen in the datagram packing PR.

hanno-arm replied:
I think we should remove all `-C "resend"` guard in all DTLS tests for now, and perhaps open an issue to add them as separate tests at a later stage; as far as I understand, if we use `reco_delay` and disable HVR's, then we shouldn't run into spurious resends.

mpg commented on library/ssl_tls.c:
```diff
@@ -2871,21 +2878,72 @@ int mbedtls_ssl_flight_transmit( mbedtls_ssl_context *ssl )
+        {
+            if( cur->next != NULL )
+            {
+                ssl->handshake->cur_msg = cur->next;
```
If `mbedtls_ssl_write_record()` returns `MBEDTLS_ERR_SSL_WANT_WRITE`, then we'll return to the application and next time the application calls a function from the SSL module, the first this this function will do is call `mbedtls_ssl_flush_output()` in order to finish sending, well before we reach the current code path again. Now the pointers in question are only used in the current function, not in `mbedtls_ssl_flush_output()`, so they can have any value we want.

However, thanks for pointing that out, because I think I had actually forgotten to exercise that in the tests, so I've now added two tests with `nbio=2` in order to make sure the code behaves correctly in this case.

k-stachowiak replied:
@mpg Thanks for the clarification. My first thought when I saw it that these pointers could have been made local, and only then I considered they may also be kept as the SSL state. Does it mean that they can be moved to the local scope here?

mpg replied:
No, I don't think they could be made local, they need to be remembered if this function is exited and re-entered later. The simple case would be if `mbedtls_ssl_write_record()` always returned zero: in that case, we would always go to the end of the loop in one call, and the loop variables could indeed be local to the function in which the loop resides.

Unfortunately, we also need to support non-blocking I/O, where `mbedtls_ssl_write_record()` might return a temporary error at any time: `MBEDTLS_ERR_SSL_WANT_WRITE`. Here our model mandates that we return control to the application, that is we need to break the loop, return from the current function, its caller, all the way up to the first function in the call stack that belongs to the application. When any function from the SSL module is called later, it needs to have enough state saved in the SSL context in order to (1) try the underlying `f_send()` call that failed temporarily, and if it succeeds (2) call again the function that was being called (eg retransmitting a flight), which will itself do the same, until we reach `mbedtls_flight_transmit()` that then needs to (3) continue the loop where it was as if nothing happened.

That's why the loop variables (in addition to some other information, such as DTLS (re)transmission state) need to be persistent (saved into the context) rather than local variables. I guess one could call this a limited form of [continuation](https://en.wikipedia.org/wiki/Continuation) embedded in the SSL context.

It took me some time to get used to it when I first started to work on the project, and even now, I'm never really confident I got it right until I passed some relevant tests with `nbio=2` (special I/O callbacks which I wrote specifically to test this). Fortunately it looks like in that case I got it right, and only forgot to add tests for it later, so thanks for the reminder!

mpg commented on library/ssl_tls.c:
```diff
@@ -2895,13 +2883,24 @@ int mbedtls_ssl_flight_transmit( mbedtls_ssl_context *ssl )
         else
         {
+            const int ret_payload = mbedtls_ssl_get_max_out_record_payload( ssl );
+            const size_t max_record_payload = (size_t) ret_payload;
```
Ok, after a small chat with Hanno, I'm not removing it after all, since in his packing work this variable is used more that once, which makes it more useful. Also, we think there is value in distinguishing between potentially-negative return values and actual sizes.

(If we could use C99-style declarations I would write the code as follows, but unfortunately we can't yet. However once compiled it should be the same.)
```c
            const int ret_payload = mbedtls_ssl_get_max_out_record_payload( ssl );

            if( ret_payload < 0 )
            {
                MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_get_max_out_record_payload",
                                          ret_payload );
                return( ret_payload );
            }

            const size_t max_record_payload = (size_t) ret_payload;
```

mpg commented:
@hanno-arm @k-stachowiak Thanks for your reviews! I believe I've addressed all of your feedback, either by changing the code, or explaining why not. Please review again.

hanno-arm commented:
@mpg I pushed commits 2c98db2 and 551835d to address the failure mentioned in the review. Could you have a look if you think they are appropriate, and potentially cherry-pick them here?

sbutcher-arm commented:
retest

mpg commented:
@hanno-arm @k-stachowiak I quick-fixed the gnutls-cli tests so that they pass also on machines where IPv6 does not work or where `localhost` resolves as IPv4 first, as discussed on skype; I also cherry-picked fixes for the SSLv3 issues as suggested by Hanno above. Please review again.

@hanno-arm Please be aware that my commit changes the `$G_CLI` default, so in some other PR of yours you're adding other tests using `$G_CLI`, you'll need to adapt them too by appending `localhost` to the command line, which was previously part of the default. This is a non-local change, so git will not show it as a conflict but it needs resolution nonetheless.

mpg commented:
Note: `tests/scripts/all.sh -k -r -m` ran successfully on the latest commit 081bd81865881b82fc5d04847189b01fe4df8c1e Travis passed on that commit too. @sbutcher-arm I'm not quite sure what the label "needs-ci" means (I know we're going to clarify that soon), but since you added it yesterday, please consider updating in light of that information.

sbutcher-arm commented:
retest

sbutcher-arm commented:
Hi @mpg. 'needs CI' means, the CI hasn't passed or is yet to run. In this instance, the CI (I think) had failed. Right now it's still failing but the build seems to be stale so I've triggered the build again.

mpg commented:
I checked the output of the "mbed TLS build & test" CI and it appears the only failing test is the timing suite, which is known to be flaky, see #1517, so is not a blocker.

mpg commented:
retest

mpg commented:
The new CI run failed again only in the timing test. I'm not triggering a new run, though feel free to do so if you feel it's useful.

mpg commented:
retest

sbutcher-arm commented on tests/ssl-opt.sh:
```diff
@@ -4877,6 +4940,693 @@ run_test    "DTLS reassembly: fragmentation, nbio (openssl server)" \
+            -C "error"
+
+# This ensures things still work after session_reset(),
+# for example it would have caught #1941.
```
We shouldn't reference GitHub issues in the test code, and especially not without explaining it's a GitHub issue. Instead if it's worth stating we should name the issue and explain it.

However - I'm not sure we have to - we should just state the test case and why we have it without mentioning bugs we've had before.

hanno-arm replied:
Addressed in https://github.com/hanno-arm/mbedtls/commit/b841b4f107aa3368ade353de8845b10858aeaee8.

sbutcher-arm commented on tests/ssl-opt.sh:
```diff
@@ -4877,6 +4940,693 @@ run_test    "DTLS reassembly: fragmentation, nbio (openssl server)" \
+## https://gitlab.com/gnutls/gnutls/issues/543
+## We can re-enable them when a fixed version fo GnuTLS is available
+## and installed in our CI system.
+##
```
I'd rather we didn't comment these tests out, but instead skipped them, so they're included in the stats of what test cases we have, but also so as 'skipped' tests.

I suggest we add a forcing function of `test_disabled` which sets `SKIP_NEXT="YES"` to force the test to be skipped. Better than commenting it out.

hanno-arm replied:
Followed your suggestion in https://github.com/hanno-arm/mbedtls/commit/3b8b40c16dbfb3df1c33025bd08c6d9bcf15a147.

sbutcher-arm commented:
retest

mpg commented:
Note to self: when creating the follow-up PR, see also https://github.com/ARMmbed/mbedtls/pull/1918#discussion_r212833922

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -2822,20 +2822,37 @@ static void ssl_swap_epochs( mbedtls_ssl_context *ssl )
  * This function must be called until state is no longer SENDING.
  */
-int mbedtls_ssl_resend( mbedtls_ssl_context *ssl )
+int mbedtls_ssl_flight_transmit( mbedtls_ssl_context *ssl )
```
@sbutcher-arm I prefer this naming scheme because it allows to broadly categorize functions according to the parts of their name. For example, there are also (pre-existing) functions `mbedtls_ssl_flight_append()` and `mbedtls_ssl_flight_free()`, and for all of them the `mbedtls_ssl_flight` prefix indicates that they're about flight handling.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -3002,31 +3127,63 @@ int mbedtls_ssl_write_record( mbedtls_ssl_context *ssl )
 
-        if( out_msg_type != MBEDTLS_SSL_HS_HELLO_REQUEST )
-            ssl->handshake->update_checksum( ssl, ssl->out_msg, len );
+        /* Update running hashes of hanshake messages seen */
```
Addressed in https://github.com/hanno-arm/mbedtls/commit/0207e533b21776e304c99b6bb9e2a2826cf2d421.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -1374,6 +1378,51 @@ void mbedtls_ssl_set_bio( mbedtls_ssl_context *ssl,
+ *                 encapsulation and encryption/authentication if any.
+ *
+ * \note           This can be called at any point during the connection, for
+ *                 example when a PMTU estimate becomes available from other
```
Addressed in https://github.com/hanno-arm/mbedtls/commit/eefe084f7299959b5138a80709aa1dc6f05f4885.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -2927,43 +3013,83 @@ void mbedtls_ssl_send_flight_completed( mbedtls_ssl_context *ssl )
+                ssl->conf->endpoint == MBEDTLS_SSL_IS_CLIENT ) )
+#endif /* MBEDTLS_SSL_PROTO_SSL3 && MBEDTLS_SSL_SRV_C */
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 1, ( "should never happen" ) );
```
I think this should be addressed as a separate issue, going through all uses of `should never happen` we have in the library.
